# Impact of progressive resistance exercise training on splicing
efficiency


## Authors

Chidimma Phoebe Echebiri, Daniel Hammarström,

## Introduction

This study is aimed at investigating splicing efficiency (SE)
differences between the young and old, and the potential impact
progressive resistance exercise could have on SE. It uses SpliceQ Melo
Costa et al. (2021) to quantify splicing efficiency in five datasets;
the \[COPD
dataset\](<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8261934/>),
\[Volume
dataset\](<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7708234/>),
\[SRP102542\](<https://pubmed.ncbi.nlm.nih.gov/28273480/>),
\[SRP043368\](<https://faseb.onlinelibrary.wiley.com/doi/epdf/10.1096/fj.14-255000>)
and \[SRP280348\](<https://pubmed.ncbi.nlm.nih.gov/33071237/>)

## Repository Organisation

`/R`

This contains the scripts with codes used for data analyses.

1.  \`SpliceQ_from_scratch.R\` details the steps in processing the
    SpliceQ data generated from RNA Seq data
2.  \`metadata_extraction.R\` shows how metadata from the Trainome files
    are generated and processed
3.  \`metadata_public_data.R\` shows how the metadata from the publicly
    available RNA-Seq data were generated  

<div id="refs" class="references csl-bib-body hanging-indent"
entry-spacing="0">

<div id="ref-demelocosta2021" class="csl-entry">

Melo Costa, Verônica R. de, Julianus Pfeuffer, Annita Louloupi, Ulf AV
Ørom, and Rosario M. Piro. 2021. “SPLICE-q: A Python Tool for
Genome-Wide Quantification of Splicing Efficiency.” *BMC Bioinformatics*
22: 1–14.

</div>

</div>
